### PR TITLE
GH-101362: Check pathlib.Path flavour compatibility at import time

### PR DESCRIPTION
**Pull request by barneygale**
This saves a comparison in `pathlib.Path.__new__()` and reduces the time taken to run `Path()` by ~5%.

* Issue: gh-101362

Automerge-Triggered-By: GH:AlexWaygood

### CODE REVIEW

File: Lib/pathlib.py [hunk 1]
```diff
@@ -709,11 +709,7 @@ def __new__(cls, *args, **kwargs):
             warnings._deprecated("pathlib.PurePath(**kwargs)", msg, remove=(3, 14))
         if cls is Path:
             cls = WindowsPath if os.name == 'nt' else PosixPath
```
Comment (AlexWaygood) [inline Lib/pathlib.py:709]: Could this condition also be calculated and stored at import time? You could have a global or class `_DEFAULT_PATH` variable, and then this condition could be changed to:

```python
if cls is Path:
    cls = _DEFAULT_PATH
```

**Comment by AlexWaygood** [comment]
Very cool idea!

File: Lib/pathlib.py (hunk 1 above)
Comment (barneygale) [reply Lib/pathlib.py:709]: Though this makes sense in theory, I can't seem to measure any significant performance impact from making this change! I might leave it, to keep the code a little simpler?

File: Lib/pathlib.py [hunk 2]
```diff
@@ -1261,9 +1257,19 @@ class PosixPath(Path, PurePosixPath):
     """
     __slots__ = ()
 
+    if os.name == 'nt':
+        def __new__(cls, *args, **kwargs):
+            raise NotImplementedError("cannot instantiate %r on your system"
+                                      % (cls.__name__,))
```
Comment (AlexWaygood) [inline Lib/pathlib.py]:
```suggestion
            raise NotImplementedError(
                f"cannot instantiate {cls.__name__!r} on your system")
```

File: Lib/pathlib.py [hunk 3]
```diff
@@ -1261,9 +1257,19 @@ class PosixPath(Path, PurePosixPath):
...
     """Path subclass for Windows systems.
 
     On a Windows system, instantiating a Path should return this object.
     """
     __slots__ = ()
+
+    if os.name != 'nt':
+        def __new__(cls, *args, **kwargs):
+            raise NotImplementedError("cannot instantiate %r on your system"
+                                      % (cls.__name__,))
```
Comment (AlexWaygood) [inline Lib/pathlib.py]:
```suggestion
            raise NotImplementedError(
                f"cannot instantiate {cls.__name__!r} on your system")
```

**Review by AlexWaygood — COMMENTED** [review]
Tiny nit: since we're touching these lines anyway, I'd prefer to modernise the code slightly by using f-strings

**Review by AlexWaygood — APPROVED** [review]
LGTM!

**Pull request merged**